File: bindings/rust/runtime/entity_gc_repository.rs
use crate::backend::{BackendResult, Key, Transaction, VersionedRecord};
use crate::runtime_entity_gc_descriptors::EntityGcDescriptor;

pub struct EntityGcEligibilityRequest {
    pub descriptor: EntityGcDescriptor,
    pub now: String,
    pub limit: usize,
}

pub struct EntityGcFinalizeRequest {
    pub descriptor: EntityGcDescriptor,
    pub key: Key,
    pub now: String,
}

pub trait EntityGcRepository<Tx: Transaction> {
    fn list_gc_eligible_tx(
        &self,
        tx: &mut Tx,
        request: &EntityGcEligibilityRequest,
    ) -> BackendResult<Vec<VersionedRecord>>;

    fn finalize_gc_tx(
        &self,
        tx: &mut Tx,
        request: &EntityGcFinalizeRequest,
    ) -> BackendResult<()>;
}

File: bindings/java/com/statespec/backend/runtime/EntityGcRepository.java
package com.statespec.backend.runtime;

import com.statespec.backend.Backend;
import java.util.List;

public interface EntityGcRepository
{
    record EligibilityRequest(
        EntityGcTypes.Descriptor descriptor,
        String now,
        int limit
    )
    {
    }

    record FinalizeRequest(
        EntityGcTypes.Descriptor descriptor,
        String key,
        String now
    )
    {
    }

    List<Backend.VersionedRecord> listGcEligibleTx(
        Backend.Transaction tx,
        EligibilityRequest request
    ) throws Backend.BackendException;

    void finalizeGcTx(
        Backend.Transaction tx,
        FinalizeRequest request
    ) throws Backend.BackendException;
}
